### PR TITLE
[opentitantool] Add JTAG support to HyperDebug backend

### PR DESCRIPTION
**Pull request by jesultra**
HyperDebug now supports the CMSIS-DAP protocol which OpenOCD already has a transport driver for.  With this change, the HyperDebug backend driver is able to start up OpenOCD with appropriate parameters to have it connect to the same HyperDebug.

### CODE REVIEW

File: sw/host/opentitanlib/BUILD [hunk 1]
```diff
@@ -304,6 +308,7 @@ rust_test(
         "src/otp/testdata/otp_ctrl_img_dev.hjson",
         "src/otp/testdata/otp_ctrl_mmap.hjson",
         "src/otp/testdata/output.vmem",
+        "//third_party/openocd:jtag_cmsis_dap_adapter_cfg",
```
Comment (pamaury) [inline sw/host/opentitanlib/BUILD:311]: Do you need to specify the data here as well? It's already specified in `:opentitanlib` I think.

File: sw/host/opentitanlib/src/util/openocd.rs [hunk 2]
```diff
@@ -25,6 +25,11 @@ use crate::util::printer;
 
 /// Represents an OpenOCD server that we can interact with.
 pub struct OpenOcdServer {
+    /// Configuration file for the particular adapter used.  If given, will override the one from
```
Comment (pamaury) [inline sw/host/opentitanlib/src/util/openocd.rs:28]: I don't know if overriding the command line one is the best thing to do as it could lead to counter-intuitive results? Other options would be to simply pass the jtag_cmsis_dap_adapter_cfg file on the command line for hyperdebug tests, or have an extra CLI argument for the hyperdebug adapter config and let the backend (cw/hyperdebug) choose which config to use?

Comment (jesultra) [reply sw/host/opentitanlib/src/util/openocd.rs:28]: I would like to work towards a solution in which the `hyperdebug` backend driver knows and passes the file that fits it, and the existing `chip_whisperer` passes the file that fits it.  That is, the backend configuration file is not passed on the command line.

I did not want to possibly disrupt the existing driver, that is why I first introduce the override field, and then next, that could become the only way to specify the OpenOCD backend configuration.

File: sw/host/opentitanlib/BUILD (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/BUILD:311]: I would think the same, and at first I had not specified, but then bazel complained.

You can even see how the CI complains that the documentation rule does not also include this same data.  I think that there is something that I am not doing right.

`in rust_doc rule //sw/host/opentitanlib:opentitanlib_doc: label '//third_party/openocd:jtag_cmsis_dap_adapter_cfg' in $(location) expression is not a declared prerequisite of this rule`
https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=129009&view=logs&jobId=3a559e2a-952e-58d2-b8db-2e604a9266d7&j=3a559e2a-952e-58d2-b8db-2e604a9266d7&t=5d859c63-56da-5106-2237-dc76198705d5

**Comment by jesultra** [comment]
By the way.  I have just verified that with the HyperDebug firmware included in this patch, I am able to use OpenOCD and gdb to debug (single step) a program running on the CW310 board, all using HyperDebug.  That is, HyperDebug can replace the Olimex JTAG dongle, and we should be able to do tests involving JTAG with the HyperDebug bitstream. This should be good preparation for running tests on the future ASIC shield, driven by HyperDebug.

**Comment by pamaury** [comment]
We have some jtag tests in CI. When this is merge, it would be good to add a hyperdebug/jtag tests as well (driven by opentitanlib). Probably it's just a matter of copying the existing one and switching the interface to hyperdebug.

File: sw/host/opentitanlib/BUILD [hunk 3]
```diff
@@ -231,6 +234,8 @@ rust_library(
         "pinmux_config": "$(location :pinmux_config)",
         "spi_passthru": "$(location :spi_passthru)",
         "hyperdebug_firmware": "$(location @hyperdebug_firmware//:hyperdebug/ec.bin)",
+        # Manual expansion of: $(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)
```
Comment (pamaury) [inline sw/host/opentitanlib/BUILD:237]: why is this necessary?

Comment (jesultra) [reply sw/host/opentitanlib/BUILD:237]: Thanks for asking.  What happened when I wrote `"openocd_cmsis_dap_adapter_cfg": "$(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)"` was that bazel complained when compiling `opentitanlib_test` that it was referring to the data dependency through `$(rootpath ...)`, but did not itself declare it.  I could make that error go away by duplicating the `data = [ ... ]` declaration in `opentitanlib_test`.  However, next thing was that bazel complained that `opentitanlib_doc` also referred to the data dependency without declaring it.  It does not seem right to duplicate the `data` that way (and `rust_doc` does not allow `data` anyway.)

If anyone knows of a better way of referring to a data files from a library like this, I am happy to make the change.

Comment (pamaury) [reply sw/host/opentitanlib/BUILD:237]: I will have a look at this.

File: sw/host/opentitanlib/src/transport/hyperdebug/mod.rs [hunk 4]
```diff
@@ -624,13 +624,30 @@ impl<T: Flavor> Transport for Hyperdebug<T> {
             Err(TransportError::UnsupportedOperation.into())
         }
     }
-
     fn jtag(&self, opts: &JtagParams) -> Result<Rc<dyn Jtag>> {
         let mut jtag = self.inner.jtag.borrow_mut();
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:628]: But it is used in the very next line!

File: sw/host/opentitanlib/BUILD (hunk 3 above)
Comment (pamaury) [reply sw/host/opentitanlib/BUILD:237]: I was able to make it work with:
```diff
diff --git a/sw/host/opentitanlib/BUILD b/sw/host/opentitanlib/BUILD
index e2b6610d1..40a93cecb 100644
--- a/sw/host/opentitanlib/BUILD
+++ b/sw/host/opentitanlib/BUILD
@@ -213,6 +213,7 @@ rust_library(
         ":pinmux_config",
         ":spi_passthru",
         "@hyperdebug_firmware//:hyperdebug/ec.bin",
+        "//third_party/openocd:jtag_cmsis_dap_adapter_cfg",
     ],
     crate_features = [
         "include_hyperdebug_firmware",
@@ -220,9 +221,6 @@ rust_library(
         "//sw/device:is_english_breakfast": ["english_breakfast"],
         "//conditions:default": [],
     }),
-    data = [
-        "//third_party/openocd:jtag_cmsis_dap_adapter_cfg",
-    ],
     proc_macro_deps = [
         "//sw/host/opentitanlib/opentitantool_derive",
         "@lowrisc_serde_annotate//annotate_derive:annotate_derive",
@@ -234,8 +232,7 @@ rust_library(
         "pinmux_config": "$(location :pinmux_config)",
         "spi_passthru": "$(location :spi_passthru)",
         "hyperdebug_firmware": "$(location @hyperdebug_firmware//:hyperdebug/ec.bin)",
-        # Manual expansion of: $(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)
-        "openocd_cmsis_dap_adapter_cfg": "external/openocd/tcl/interface/jtag_cmsis_dap_adapter_cfg",
+        "openocd_cmsis_dap_adapter_cfg": "$(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)",
     },
     deps = [
         "//hw/top_earlgrey/sw/autogen/chip:top_earlgrey",
```
I think the problem is that you put it in `data` instead of `compile_data`.

Comment (jesultra) [reply sw/host/opentitanlib/BUILD:237]: What do you mean by "make it work".  Did you attempt running the code and see opentitantool being able to spawn openocd, and openocd accessing the file?

As far as I understand, `compile_data`, as its name suggests, is only available at compile time, and not later at runtime, when this file is going to be needed for an openocd sub-process.  (Maybe I could do something funny, and include the data in the binary by `include_str!()`, and start the openocd sub-process with something like `/dev/fd/17` on the command line as the configuration file, and create anonymous pipes to send the data through.  This might work, but it is also possible that openocd wants to `seek()` in the file, or do something else that does not work on pipes.  And I do not know if the Rust sub-process library allows the creation of such anonymous pipes besides stdin, stdout and stderr.  I do not want to go there.)

Comment (pamaury) [reply sw/host/opentitanlib/BUILD:237]: Ah sorry, I misunderstood what you are trying to achieve here. You are trying to embed the **path** (not content) to a bazel-managed file when you create the library. I should point out that this is an extremely fragile way of doing things because you are relying on a stable path at compile and execution time. For this you need to pass the file as both compile_data and data. Then you ca use rootpath that actually guarantees to use a workspace-relative path so it will be the same path at compile time and execution time.

```diff
diff --git a/sw/device/tests/BUILD b/sw/device/tests/BUILD
index ad2241c8c..f42eb7b56 100644
--- a/sw/device/tests/BUILD
+++ b/sw/device/tests/BUILD
@@ -2827,6 +2827,26 @@ opentitan_functest(
     ],
 )
 
+opentitan_functest(
+    name = "openocd_hyperdebug_test",
+    srcs = ["example_test_from_flash.c"],
+    cw310 = cw310_params(
+        interface = "hyper310",
+        bitstream = "//hw/bitstream/hyperdebug:rom_with_fake_keys_otp_test_unlocked0",
+        test_cmds = [
+            "--bitstream=\"$(rootpath {bitstream})\"",
+            "--bootstrap=\"$(rootpath {flash})\"",
+        ] + OPENTITANTOOL_OPENOCD_TEST_CMDS,
+    ),
+    data = OPENTITANTOOL_OPENOCD_DATA_DEPS,
+    targets = ["cw310_rom_with_fake_keys"],
+    test_harness = "//sw/host/tests/chip/jtag:openocd_test",
+    deps = [
+        "//sw/device/lib/runtime:log",
+        "//sw/device/lib/testing/test_framework:ottf_main",
+    ],
+)
+
 opentitan_functest(
     name = "spi_device_ottf_console_test",
     srcs = ["spi_device_ottf_console_test.c"],
diff --git a/sw/host/opentitanlib/BUILD b/sw/host/opentitanlib/BUILD
index e2b6610d1..c6870b338 100644
--- a/sw/host/opentitanlib/BUILD
+++ b/sw/host/opentitanlib/BUILD
@@ -213,6 +213,7 @@ rust_library(
         ":pinmux_config",
         ":spi_passthru",
         "@hyperdebug_firmware//:hyperdebug/ec.bin",
+        "//third_party/openocd:jtag_cmsis_dap_adapter_cfg",
     ],
     crate_features = [
         "include_hyperdebug_firmware",
@@ -234,8 +235,12 @@ rust_library(
         "pinmux_config": "$(location :pinmux_config)",
         "spi_passthru": "$(location :spi_passthru)",
         "hyperdebug_firmware": "$(location @hyperdebug_firmware//:hyperdebug/ec.bin)",
-        # Manual expansion of: $(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)
-        "openocd_cmsis_dap_adapter_cfg": "external/openocd/tcl/interface/jtag_cmsis_dap_adapter_cfg",
+        # NOTE: here we must use rootpath, this is because we are using file as both compile_data
+        # to embed the path name, and as data to actually load it at runtime. For this purpose,
+        # the name must be independent of the execution environment's path. This guarantee is provided
+        # by rootpath that will generate a pure workspace-relative path:
+        # https://bazel.build/reference/be/make-variables#predefined_label_variables
+        "openocd_cmsis_dap_adapter_cfg": "$(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)"
     },
     deps = [
         "//hw/top_earlgrey/sw/autogen/chip:top_earlgrey",
```
I actually tried it with
```bash
./bazelisk.sh test --test_output=streamed //sw/device/tests:openocd_hyperdebug_test
```
and it worked.

Comment (jesultra) [reply sw/host/opentitanlib/BUILD]: Thank you, this helped immensely.

I have done the changes as you have suggested.  I have also added another CL into this PR, which refactors how configuration files for the existing Olimex JTAG adapter is passed into Rust code.  I think it is a simplification, let me know what you think.

I could back out the refactor, and instead add one more JTAG command line argument for the HyperDebug adapter, though it does not feel right to me that the caller of `opentitantool`, who already specify `--interface cw310` or `--interface hyper310` to have to also know details about which OpenOCD configuration file each of the backends need.

Comment (jwnrt) [reply sw/host/opentitanlib/BUILD]: I agree with trying to keep opentitanlib separate from Bazel at runtime, and embedding the config sounds sensible. Writing it to a temporary file sounds fiddly, but maybe there’s a way to pass config through the CLI?

Comment (jesultra) [reply sw/host/opentitanlib/BUILD]: I think you can pass the entire content of openocd config files on the command line with the `-c` switch, instead of the `-f` switch.  Let me look into that.

I agree that it could make sense to swap out the olimex configuration file, and hence that it should be passed as a parameter.  I do not think that it belongs in the `io::jtag::JtagParams`, which is part of the `Jtag` trait that is shared by all backend drivers, though, but rather should be a parameter to the `ChipWhisperer` constructor.  I am going to attempt such a refactor as well.

How about the openocd executable itself?  We currently depend on a patched version.  Should that be embedded into opentitanlib as well, so that at runtime it could write a temporary file, make it executable, and start it?  For "outside" users who do not use bazel, it probably makes sense that opentitantool attempts to use the local openocd from the users $PATH, (as long as we know that there is a chance it could work.)  Does anyone know the nature of the local changes to openocd? Could they be upstreamed such that we do not need to depend on patches?

**Comment by jesultra** [comment]
Ok, I have changed this PR such that the path to `openocd` is still passed via a field in `JtagParams`, for all backends.  The path to the Olimex adapter config is now passed via a field in `ChipWhispererOpts`, as it currently does not apply to other backends.
The other OpenOCD config files, the ones about OpenTitan JTAG TAPs, and the one for the CMSIS-DAP compatible adapters (HyperDebug), are compiled into opentitanlib, and passed as strings on the command line to `openocd` via the `-c` switch.  (The combined command line probably exceeds a kilobyte.)

The new HyperDebug firmware included in this PR seems to trigger a bug in the I2C backend transport.  I will look at that Monday, but I feel that the JTAG changes in this CL is fully ready for review.

File: sw/host/opentitanlib/src/backend/chip_whisperer.rs [hunk 5]
```diff
@@ -15,6 +15,11 @@ pub struct ChipWhispererOpts {
     /// Comma-separated list of Chip Whisperer board UARTs for non-udev environments. List the console uart first.
     #[arg(long, alias = "cw310-uarts")]
     pub uarts: Option<String>,
+
+    /// Path to OpenOCD JTAG adapter config file to use (usually Olimex, but could be another
+    /// adapter).
+    #[arg(long)]
+    pub openocd_adapter_config: Option<String>,
```
Comment (pamaury) [inline sw/host/opentitanlib/src/backend/chip_whisperer.rs:22]: I think it should probably be a `Option<PathBuf>`, I am not sure why we defined the paths as `Option<String>` when we did the openocd backend. Can you try and see if that causes any issue?

**Comment by pamaury** [comment]
Thanks @jesultra I think it's much better this way. If the command-line becomes a problem, we might look into creating a temporary file to pass it to OpenOCD.

**Review by pamaury — APPROVED** [review]

File: sw/host/opentitanlib/src/backend/chip_whisperer.rs (hunk 5 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/backend/chip_whisperer.rs]: I have changed the parameter into `Option<PathBuf>`.  I have also added a small refactor CL to this PR, which eliminates the use of  `Vec<&str>` in the starting of the OpenOCD sub-process, in order to allow some of the arguments to be `&PathBuf` while others remain `&str`.
